homeassistant/components/husqvarna_automower/__init__.py

```diff
@@ -0,0 +1,100 @@
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+from homeassistant.helpers.issue_registry import IssueSeverity, async_create_issue
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN, PLATFORMS
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AutomowerDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching Husqvarna data."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize data updater."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+        )
+        api_key = None
+        ap_storage: dict = cast(dict, hass.data.get("application_credentials"))
+        ap_storage_data: dict = ap_storage[DATA_STORAGE].__dict__["data"]
+        for k in ap_storage_data:
+            api_key = ap_storage_data[k]["client_id"]
+        entry_dict = entry.as_dict()
+        access_token = entry_dict["data"]["token"]
+        scope = entry_dict["data"]["token"]["scope"]
+        if "amc:api" not in scope:
+            async_create_issue(
+                hass,
+                DOMAIN,
+                "wrong_scope",
+                is_fixable=False,
+                severity=IssueSeverity.WARNING,
+                translation_key="wrong_scope",
+            )
+        low_energy = False
+        self.session = aioautomower.AutomowerSession(api_key, access_token, low_energy)
+        self.session.register_token_callback(
+            lambda token: hass.config_entries.async_update_entry(
+                entry,
+                data={"auth_implementation": DOMAIN, CONF_TOKEN: token},
+            )
+        )
+
+    async def _async_update_data(self) -> None:
+        """Fetch data from Husqvarna."""
+        try:
+            await self.session.connect()
+        except AsyncioTimeoutError as error:
+            _LOGGER.debug("Asyncio timeout: %s", error)
+            raise ConfigEntryNotReady from error
+        except Exception as error:
+            _LOGGER.debug("Exception in async_setup_entry: %s", error)
+            raise ConfigEntryAuthFailed from Exception
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up this integration using UI."""
+    if hass.data.get(DOMAIN) is None:
+        hass.data.setdefault(DOMAIN, {})
+    coordinator = AutomowerDataUpdateCoordinator(
+        hass,
+        entry=entry,
+    )
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Handle unload of an entry."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    with contextlib.suppress(Exception):
```

Comment:
What could raise?

Reply:
If there is for example no internet there could be an error.
Or if the session is broken out of any reason, there would be no possibility to unload the integration.
So there could be nearly anything rising.

Reply:
The library should handle transport errors and  raise library specific errors instead. When using aiohttp as a transport only the aiohttp exceptions will be relevant to catch in the transport layer.

Reply:
> What could raise?

Basically everything. So this just ensures, that the entry is unloaded, whatever happens.